src/Microsoft.CodeAnalysis.Razor.Workspaces/WorkspaceMonitor.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    public class WorkspaceMonitor
+    {
+        private readonly Workspace _workspace;
+
+        public WorkspaceMonitor(Workspace workspace)
+        {
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            _workspace = workspace;
+
+            _workspace.WorkspaceChanged += OnWorkspaceChanged;
```

Comment:
You will need to unsubscribe at some point. If you ever define an event you either need the *source* and *sink* to have the same lifetime, or you need to unsubscription protocol.